### DISCUSSION TITLE
Support for jinja alternatives

### DISCUSSION THREAD

**Discussion by azmeuk** [pallets/flask#5465] [Ideas]
I suppose this have low chances to happen, but that might be fun to discuss though.
I stumbled upon Armin Ronacher's [minijinja](https://github.com/mitsuhiko/minijinja/tree/main/minijinja-py) project, that is a rewrite in Rust. The project announce to be kinda compatible with jinja

> MiniJinja attempts a reasonably high level of compatibility with Jinja2, but it does not try to achieve this at all costs

I have noticed that jinja is sometimes the performance bottleneck, notably in unit tests, so the Rust rewrite looks appealing.
Flask has a hard dependency towards jinja at the moment, but that would be awesome if it provided some kind of interface that could be filled either by jinja or minijinja.
I suppose this would require a lot of work making the code generic, and I am not sure what would be the performance benefit in production environments, but that would be cool :)

What do you think?

**Comment by davidism** [comment]
I have actually been thinking about this for a long time, especially while I was doing work last year to move things like `redirect` and `url_for` to be methods on the Flask app rather than global functions. It would be nice to allow using different template engines, but I'm not sure if it's realistic. There's no standard API for accessing/rendering templates, so right now everything assumes Jinja. There's also no standardization on features or syntax, so extensions that expect Jinja will fail if Mako is plugged in, and I can't think of a good way around that either. Even within different implementations of Jinja, the language itself isn't standardized, so I can't guarantee that they'll all be able to render each other's templates.

So yes, I am interested in this, but there's no active work on it.

Additionally, if speed in Jinja is an issue, you should look into how you have the caches configured, and look into precompiling the templates. I also have vague ideas around improving the spec, parser, and compiler which could improve things in the long term.

**Reply by azmeuk** [reply]
> Additionally, if speed in Jinja is an issue, you should look into how you have the caches configured

Wow, thank you for that. I did not know jinja caches was a thing. On several testsuites I could save approximately 30% of execution time with [a few lines of code](https://github.com/spiral-project/ihatemoney/pull/1305/files).
